servicetalk-opentelemetry-http/src/main/java/io/servicetalk/opentelemetry/http/GrpcSemanticAttributesExtractor.java

```diff
@@ -0,0 +1,115 @@
+
+    static final AttributeKey<String> NETWORK_PEER_ADDRESS = AttributeKey.stringKey("network.peer.address");
+    static final AttributeKey<Long> NETWORK_PEER_PORT = AttributeKey.longKey("network.peer.port");
+
+    static final GrpcSemanticAttributesExtractor INSTANCE = new GrpcSemanticAttributesExtractor();
+
+    private GrpcSemanticAttributesExtractor() {
+    }
+
+    @Override
+    public void onStart(AttributesBuilder attributes, Context parentContext, RequestInfo request) {
+        // Set the RPC system to gRPC
+        attributes.put(RPC_SYSTEM, "grpc");
+
+        // Extract service and method from gRPC path: /service.name/MethodName
+        String path = request.getMetadata().path();
+        extractServiceAndMethod(path, attributes);
+        extractPeerAddress(request, attributes);
+    }
+
+    @Override
+    public void onEnd(AttributesBuilder attributes, Context context, RequestInfo request,
+                      @Nullable GrpcTelemetryStatus response, @Nullable Throwable error) {
+        // Extract gRPC status code from response headers/trailers (trailers first per gRPC spec)
+        if (response != null) {
+            extractGrpcStatusCode(response, attributes);
+        }
+    }
+
+    private static void extractServiceAndMethod(String path, AttributesBuilder attributes) {
+        if (path == null || path.isEmpty()) {
+            return;
+        }
+
+        // gRPC path format: /service.name/MethodName
+        // Remove leading slash if present
+        String normalizedPath = path.startsWith("/") ? path.substring(1) : path;
+
+        int lastSlashIndex = normalizedPath.lastIndexOf('/');
+        if (lastSlashIndex > 0) {
+            String service = normalizedPath.substring(0, lastSlashIndex);
+            String method = normalizedPath.substring(lastSlashIndex + 1);
+
+            attributes.put(RPC_SERVICE, service);
+            attributes.put(RPC_METHOD, method);
+        }
+    }
+
+    private static void extractGrpcStatusCode(GrpcTelemetryStatus response, AttributesBuilder attributes) {
+        if (response.hasGrpcStatusCode()) {
+            long grpcStatusCode = response.getGrpcStatusCode();
+            attributes.put(RPC_GRPC_STATUS_CODE, grpcStatusCode);
+        }
+    }
+
+    private static void extractPeerAddress(RequestInfo requestInfo, AttributesBuilder attributes) {
+        ConnectionInfo connectionInfo = requestInfo.getConnectionInfo();
+        if (connectionInfo != null) {
+            SocketAddress peerResolvedAddress = connectionInfo.remoteAddress();
+            if (peerResolvedAddress instanceof InetSocketAddress) {
+                InetSocketAddress inetSocketAddress = (InetSocketAddress) peerResolvedAddress;
+                attributes.put(
+                        NETWORK_PEER_ADDRESS, inetSocketAddress.getAddress().getHostAddress());
+                attributes.put(NETWORK_PEER_PORT, inetSocketAddress.getPort());
```

Comment:
We can also extract [network.transport](https://opentelemetry.io/docs/specs/semconv/registry/attributes/network/) and [network.type](https://opentelemetry.io/docs/specs/semconv/registry/attributes/network/) in this case.

Reply:
I left `network.transport` out deliberately since it seem it will essentially always be `tcp`. I'm less confident about `network.type`: does it add enough value to justify its overhead?